api/routers/chat.py

```diff
@@ -10,73 +10,97 @@
-            honcho=honcho,
-        ).stream()
-        thought = ""
-        for chunk in thought_stream:
-            thought += chunk
-            yield chunk
+        def convo_turn():
+            thought_stream = ThinkCall(
+                user_input=inp.message,
+                app_id=app.id,
+                user_id=user.id,
+                session_id=str(inp.conversation_id),
+                honcho=honcho,
+            ).stream()
+            thought = ""
+            for chunk in thought_stream:
+                thought += chunk
+                yield chunk
 
-        yield "❀"
-        response_stream = RespondCall(
-            user_input=inp.message,
-            thought=thought,
-            app_id=app.id,
-            user_id=user.id,
-            session_id=str(inp.conversation_id),
-            honcho=honcho,
-        ).stream()
-        response = ""
-        for chunk in response_stream:
-            response += chunk
-            yield chunk
-        yield "❀"
+            yield "❀"
+            response_stream = RespondCall(
+                user_input=inp.message,
+                thought=thought,
+                app_id=app.id,
+                user_id=user.id,
+                session_id=str(inp.conversation_id),
+                honcho=honcho,
+            ).stream()
+            response = ""
+            for chunk in response_stream:
+                response += chunk
+                yield chunk
+            yield "❀"
 
-        honcho.apps.users.sessions.messages.create(
-            is_user=True,
-            session_id=str(inp.conversation_id),
-            app_id=app.id,
-            user_id=user.id,
-            content=inp.message,
-        )
-        new_ai_message = honcho.apps.users.sessions.messages.create(
-            is_user=False,
-            session_id=str(inp.conversation_id),
-            app_id=app.id,
-            user_id=user.id,
-            content=response,
-        )
-        honcho.apps.users.sessions.metamessages.create(
+            honcho.apps.users.sessions.messages.create(
+                is_user=True,
+                session_id=str(inp.conversation_id),
+                app_id=app.id,
+                user_id=user.id,
+                content=inp.message,
+            )
+            new_ai_message = honcho.apps.users.sessions.messages.create(
+                is_user=False,
+                session_id=str(inp.conversation_id),
+                app_id=app.id,
+                user_id=user.id,
+                content=response,
+            )
+            honcho.apps.users.sessions.metamessages.create(
+                app_id=app.id,
+                session_id=str(inp.conversation_id),
+                user_id=user.id,
+                message_id=new_ai_message.id,
+                metamessage_type="thought",
+                content=thought,
+            )
+
+        return StreamingResponse(convo_turn())
+    except Exception as e:
```

Comment:
Might just be I don't know how these work, but is there a change that one of these errors could occur after the generator is finished and the messages and metamessages are created? 

Unclear on the relationship between the try catch block and the Streaming Response method . If there is an error in the middle of the stream what happens?

Reply:
Very good questions! My [latest commit](https://github.com/plastic-labs/tutor-gpt/pull/151/commits/c044b58531ec57b51cee61a930f4d0784c9b29cc) employs a background process for the honcho calls to separately handle/log any potential errors while (ideally) not interfering with the `StreamingResponse`. I think in the case of say a network interruption or rate limit error, it would hit this exception and return an error to the client which would (read: should) handle this interruption gracefully - see [more](https://gist.github.com/bLopata/8a7a1dd45818d9cad7a9fd0fd60113a8).

Reply:
My worry with this approach is that if there is an error while saving a message to honcho then that is not propagated to the front-end / user. Meaning it will look like their message sent and the conversation is fine, but if they reload the messages will be gone without any indication of an error occurring. 

It might make sense to make separate try catch blocks or separate Exceptions for the different types of errors that the LLM vs the honcho calls are making and report them to the user differently, without using the background tasks